src/examples/algorithms/AnomalyDetection.cpp

```diff
@@ -0,0 +1,194 @@
+ using nupic.core (pure C++). Its purpose is to serve as a working
+ example similar to hotgym from python repo. 
+
+ The file is separated into 2 parts: 
+ :AnomalyDetection class: 
+   Provides the HTM anomaly detection functionality; 
+   parametric constructor lets us set the most useful parameters.
+   compute() which returns anomaly score for given input.
+ :main():
+   Responsible for parsing CSV input files, calling an AnomalyDetection
+   instance and writing results to a file. 
+   Optionally se DEBUG_LEVEL to see desired details. 
+   Available is also functionality to benchmark performance of the anomaly
+   detection using provided stopwatch Timer. 
+
+ Usage:
+   run $RELEASE/bin/anomaly_example 
+ */
+
+#include <iostream>
+#include <fstream> // for file output
+#include <vector>
+#include <string>
+#include <cstdlib>
+
+#include <csv.h> // external CSV parser
+
+#include "nupic/encoders/ScalarEncoder.hpp"
+#include "nupic/algorithms/SpatialPooler.hpp"
+#include "nupic/algorithms/Cells4.hpp"
+#include "nupic/algorithms/Anomaly.hpp"
+#include "nupic/os/Timer.hpp"
+#include "nupic/utils/VectorHelpers.hpp"
+
+using namespace std; // generic namespaces are included here
+using namespace nupic;
+using namespace nupic::utils;
+using namespace nupic::algorithms::spatial_pooler;
+using namespace nupic::algorithms::Cells4;
+using namespace nupic::algorithms::anomaly;
+
+const int DEBUG_LEVEL =1; //0=no debug (also disabled timer), ..
+
+class AnomalyDetection
+{
+
+  private:
+    ScalarEncoder encoder;
+    SpatialPooler sp;
+    Cells4 tp;
+    Anomaly anomaly; 
+    std::vector<UInt> lastTPOutput_;
+
+  public:
+    Real compute(Real x) {
+        auto scalar = std::vector<Real32>(encoder.getOutputWidth());
+        encoder.encodeIntoArray(x, scalar.data());
+
+        // For some reason, the ScalarEncoder outputs a Real32 vector,
+        // which is incomaptible with the SpatialPooler's input, hence a new
+        // vector is created with the required type, and the elements
+        // are converted.
+        auto uint_scalar = VectorHelpers::castVectorType<Real32, UInt>(scalar);
+        if (DEBUG_LEVEL > 1) {
+          std::cout << "Scalar encoder: ";
+          VectorHelpers::print_vector(uint_scalar);
+        }
+
+        std::vector<UInt> spOutput(sp.getNumColumns());
+        sp.compute(uint_scalar.data(), true, spOutput.data());
+        sp.stripUnlearnedColumns(spOutput.data());
+        if (DEBUG_LEVEL > 2) {
+          std::cout << "Spatial pooler: ";
+          VectorHelpers::print_vector(VectorHelpers::binaryToSparse<UInt>(spOutput), ",");
+          std::cout << std::endl;
+        }
+
+        // The temporal pooler uses Real32, so the vector must be converted again
+        auto tpInput = VectorHelpers::castVectorType<UInt, Real>(spOutput);
+        std::vector<Real> tpOutput(tp.nCells());
+        tp.compute(tpInput.data(), tpOutput.data(), true, true);
```

Comment:
@BoltzmannBrain ...this is a bit confusing; `tm.compute()` returns `void` and neither takes any output-vector as a param, so I'm guessing the "result" is in the state of the TM (wasn't the Py-TM implemented as functionaly, i.e state-less?), so the result must be accessed by some `tm.getXXX()`? 

And then, TP has param `outputMode`, which lets you set what cells propagate to the output (active, predicted, both(default)). Am I supposed to do that manually with TM? Would be nice to unify the behavior.